homeassistant/components/unifiprotect/data.py

```diff
@@ -188,3 +214,40 @@ def async_signal_device_id_update(self, device_id: str) -> None:
         _LOGGER.debug("Updating device: %s", device_id)
         for update_callback in self._subscriptions[device_id]:
             update_callback()
+
+    @callback
+    def _async_get_device_from_ref(
+        self, ref: ProtectDeviceRef
+    ) -> ProtectAdoptableDeviceModel | NVR | None:
+
+        if ref.model == ModelType.NVR:
+            if ref.device_id == self.api.bootstrap.nvr.id:
+                return self.api.bootstrap.nvr
+            return None  # pragma: no cover
+
+        device = getattr(self.api.bootstrap, f"{ref.model}s").get(ref.device_id)
+        return cast(Optional[ProtectAdoptableDeviceModel], device)
+
+    @callback
+    def async_get_ufp_device_from_device(
+        self, device_id: str
+    ) -> ProtectAdoptableDeviceModel | NVR | None:
+        """Get UniFi Protect device from HA device ID."""
+
+        ref = self._device_to_device_map.get(device_id)
+        if ref is None:
+            return None  # pragma: no cover
+
+        return self._async_get_device_from_ref(ref)
+
+    @callback
+    def async_get_ufp_device_from_entity(
+        self, unique_id: str
+    ) -> ProtectAdoptableDeviceModel | NVR | None:
+        """Get UniFi Protect device from HA entity unique ID."""
+
+        ref = self._entity_to_device_map.get(unique_id)
+        if ref is None:
+            return None  # pragma: no cover
```

Comment:
```suggestion
        ref = self._entity_to_device_map[unique_id]
```

https://github.com/home-assistant/core/pull/72212/files#r877528417